sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -201,15 +206,20 @@ public synchronized boolean advance() throws IOException {
     private synchronized boolean readNextRecord() throws IOException {
       while (decoder == null || decoder.isEnd()) {
         if (!responseIterator.hasNext()) {
+          fractionConsumed = 1d;
           return false;
         }
 
+        // N.B.: For simplicity, we update fractionConsumed once a new response is fetched, not
+        // when we reach the end of the current response. In practice, this choice is not
+        // consequential.
+        fractionConsumed = fractionConsumedFromLastResponse;
```

Comment:
We should only update the fraction consumed after this record is returned through an getCurrent() call. Otherwise this record has not been consumed yet. For example, fraction consumed should not become 1.0 till the runner gets the final record.

Reply:
I was under the impression that getCurrent() should not modify any state.

The JavaDoc doesn't explicitly call out where the progress should be updated, but it mentions, "Should return 1 after [...] Source.Reader.advance() call that returns false."[1] Reading this and the JavaDoc for advance() and getCurrent() leads me to believe that advance() is the appropriate place to update progress.

Please note that in the code I have written, progress is only updated after you have called advance() enough times to go through all rows in the response that returned the new progress. You can see an example of this in the unit tests.

I also found two examples where the progress is updated in advance():
 
- https://github.com/apache/beam/blob/c2f0d282337f3ae0196a7717712396a5a41fdde1/sdks/java/io/hadoop-format/src/main/java/org/apache/beam/sdk/io/hadoop/format/HadoopFormatIO.java#L838

- https://github.com/apache/beam/blob/c2f0d282337f3ae0196a7717712396a5a41fdde1/sdks/java/io/hbase/src/main/java/org/apache/beam/sdk/io/hbase/HBaseIO.java#L449 (uses range tracker that updates the progress; note that getCurrent() does not modify any state)

[1] https://beam.apache.org/releases/javadoc/2.13.0/org/apache/beam/sdk/io/BoundedSource.BoundedReader.html#getFractionConsumed--

Reply:
Thanks for clarifying. Current implementation (that updates the progress for a record in the advance() call following the getCurrent() call that returns the record) seems to be correct.